### ISSUE TITLE
Naive Bayes model: status and next steps

### ISSUE THREAD

**Issue by tarrade** [tarrade/proj_NLP_text_classification_with_GCP#42]
Some information about few test I did:
- training is now done on GCP using scikit-learn 0.22 (check the setup file)
- i used the naive model without pipeline
- i used the count vectorized
- i trained on  2,483,085 event ("--frac 0.1") for 7h48 for an accuracy of 0.69 % on the test set
https://console.cloud.google.com/ai-platform/jobs/stackoverflow_sklearn_20191209_125532/charts/cpu?authuser=0&cloudshell=false&project=nlp-text-classification
- I did hyperparameter tunining on 248,599 events ("--frac 0.01") for 1h16 with the best accuray of 0.71%
https://console.cloud.google.com/ai-platform/jobs/stackoverflow_sklearn_20191211_094241/charts/cpu?authuser=0&cloudshell=false&project=nlp-text-classification

![image](https://user-images.githubusercontent.com/12021701/70692019-c7cfca80-1cba-11ea-8b72-c48be66d3eb4.png)

Conclusion:
- this model is not very good (it was a quick test)
- adding more data doesn't help
- since it run on one CPU this is very slow
- no issue with memory
- didn't check memory leak

Next
- try to find model that we can trained on multiple CPU
- use pipeline
- check memory leak
- train model first on small subset of the data and tried later on few millions of events
- check confusion matrix to understand for which category we have issue
- check a few example of wrongly classifier event to understand better the issue

**Comment by gprinz** [comment]
Thank you @tarrade for providing this overview. I see the following problems:

- Its quite unclear to what extent we can save processing time when parallelizing the training process.  I think, the model takes more than 1 hour to train due to the hyperparameter tuning and not the mdoel fitting. Parallelizing Bayesian optimization is not possible with our setup. Of course, we can parallelize the model fitting when selecting an appropriate model, but we will not save much time.

- Introducing parallel processing brings us back to the problem of a high memory load.

However, with this argumentation, we will make no progress. Thus, I will take a look at your points today and try to find the best solution for us.

**Comment by tarrade** [comment]
Before
- couldn't run process more than 200'00 with a machine with 200 GB memory

Now
- could process 2.5 M event on a 8 GB memory machine but the model is taking 8h to run

Before doing hyper parameter tuning we need to see how we can process more event. Right now current model use only one CPU. Finding model that can use multiple CPU will speep the training that a huge amout.

Concerning hyperparameter tuning, I think it is better to not do it on a single machine like it is done with scikit-learn. AI Platform allow us to do that in a better way.

What is the issue with the memory you are having ? training only or training+hyperparamater tuning ?

**Comment by gprinz** [comment]
With the SGDclassifier, there is no parallelization despite the parameter **n_jobs** set to -1.

![image](https://user-images.githubusercontent.com/32190139/70806464-aacef080-1dbb-11ea-8f96-9e184fe79317.png)

For me, it looks like a bug that must be investigated in detail:
https://github.com/scikit-learn/scikit-learn/issues/15882

When proceeding with Random Forest, parallelization works well.
![image](https://user-images.githubusercontent.com/32190139/70810922-f1751880-1dc4-11ea-9a38-01d8815adc56.png)

In the case above, the number of trees has been set to 10. Accordingly, there are only 10 parallel processes (%CPU = 1'000). Parallelization is not possible within a single tree.

**Next steps**
Make a better module and run it on the AI Platform.

**Comment by tarrade** [comment]
Conclusion:
poor documentation on scikiet-learn about multi core capabilities
- Naive Bayes model: cannot run on more than 1 CPU -> useless for us: **not an option**
- SGDclassifier model: from the discussion it seems to do a poor job  -> useless for us: **not an option**

-> only option I found is Random Forest and you did the implementation

Let use that for the next steps. Honesty I was not thinking that we will be limited by that.

My conclusion is that anything that takes more than 4 hours to runs with scikit-learn, it is much to use better to use Deep Learning like TF:

1. sub task can run on paralell
2. can run on cluster
3. can use accelerator like GPU or TPU
4. is design to run with BigData

scikit-learn is nice but for small dataset <<1M events (depend of the number of features)
